File: code/library/src/main/java/org/fbme/lib/iec61499/declarations/AlgorithmBody.java
package org.fbme.lib.iec61499.declarations;

import org.fbme.lib.common.ContainedElement;
import org.fbme.lib.st.statements.Statement;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface AlgorithmBody extends ContainedElement {

    AlgorithmLanguage<?> getLanguage();

    AlgorithmDeclaration getContainer();

    interface ST extends AlgorithmBody {

        @Override
        default AlgorithmLanguage<ST> getLanguage() {
            return AlgorithmLanguage.ST;
        }

        List<Statement> getStatements();
    }

    interface Unknown extends AlgorithmBody {

        @NotNull String getText();

        void setText(@NotNull String text);
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/declarations/AlgorithmBody.kt
package org.fbme.lib.iec61499.declarations

import org.fbme.lib.common.ContainedElement
import org.fbme.lib.st.statements.Statement

interface AlgorithmBody : ContainedElement {
    val language: AlgorithmLanguage<*>
    override val container: AlgorithmDeclaration?

    interface ST : AlgorithmBody {
        override val language: AlgorithmLanguage<ST>
            get() {
                return AlgorithmLanguage.ST
            }

        val statements: MutableList<Statement>
    }

    interface Unknown : AlgorithmBody {
        var text: String
    }
}
